docs/Personalities.md

```diff
@@ -0,0 +1,276 @@
+   certificates.)
+
+### Canonicalization
+
+A personality may also perform **canonicalization** on incoming data, to
+convert equivalent formulations of the same underlying data to a single
+canonical format, avoiding needless duplication. For example, keys in JSON
+dictionaries could be sorted, or Unicode string data could be normalised.
+
+One particularly important form of canonicalization is **de-duplication** of
+timestamped data: if the Merkle leaf contents includes a timestamp indicating
+when the leaf was added, subsequent submissions of the same data with a new
+timestamp should not be treated as a new leaf.
+
+This use case is supported by the Trillian API, which includes the concept of
+an *identity hash* associated with a data item: items with the same identity
+hash are assumed equal, even if the full Merkle leaf structure differs
+(e.g. because it has a different timestamp attached).
+
+This means that a personality which wants to prevent duplicates entries can
+do so by setting the `LeafIdentityHash` value on new Merkle tree leaves
+appropriately. The personality code also needs to cope with duplicates: the
+Merkle tree leaf in a successful `QueueLeaves` response may not be the same
+as the tree leaf in the corresponding `QueueLeaves` request (it will have the
+same `LeafIdentityHash` but may differ in fields that are not covered by this
+hash, e.g. it may have an earlier timestamp).
+
+
+### Auditability
+
+If the personality and the Trillian core services are maintained by different
+operators, then there may be an implicit trust boundary between the two at the
+Trillian gRPC interface.
+
+If this is the case, then the personality may need to be responsible for
+storing data that allows auditing of that trust boundary.
+
+For example, if an external monitor detects that a signed tree head is not
+consistent with an earlier signed tree head, is there enough information
+available to determine whether this is a problem with the personality or with
+the core Trillian service?
+
+If the personality maintains a store of the signed log roots provided by the
+Trillian core, it can then use this to audit failure cases and assign blame
+appropriately &ndash; effectively acting as a monitor for the Trillian core
+service.
+
+
+## Practical Responsibilities
+
+### External API
+
+Some transparency scenarios may have the personality code embedded into
+another application, but in many environments the personality provides an
+externally-visible API.
+
+The design of such an API depends on the application, but key points to
+consider include:
+
+ - **syntax** &ndash; the mechanism used to access the API, where possibilities
+   include:
+   - HTTP(S) + JSON: A web-based interface has the advantage of being
+     accessible using standard tools, ranging from
+     [standard](https://golang.org/pkg/net/http/)
+     [libraries](https://docs.python.org/3/library/http.client.html) to
```

Comment:
nit: I'm not a big fan of individual words linking to different things, as it's not always obvious. Perhaps add "(e.g. [Go](...) and [Python](...) http packages)".

Reply:
Personally, I like using different words for different links to demonstrate a range of things.

Reply:
:+1: